src/nvim/lua/executor.c

```diff
@@ -0,0 +1,576 @@
+    if (s == NULL) {
+      PRINT_ERROR(
+          "<Unknown error: lua_tolstring returned NULL for tostring result>");
+    }
+    ga_concat_len(&msg_ga, s, len);
+    if (curargidx < nargs) {
+      ga_append(&msg_ga, ' ');
+    }
+    lua_pop(lstate, 1);
+  }
+#undef PRINT_ERROR
+  lua_pop(lstate, nargs + 1);
+  ga_append(&msg_ga, NUL);
+  {
+    const size_t len = (size_t)msg_ga.ga_len - 1;
+    char *const str = (char *)msg_ga.ga_data;
+
+    for (size_t i = 0; i < len;) {
+      const size_t start = i;
+      while (i < len) {
+        switch (str[i]) {
+          case NUL: {
+            str[i] = NL;
+            i++;
+            continue;
+          }
+          case NL: {
+            str[i] = NUL;
+            i++;
+            break;
+          }
+          default: {
+            i++;
+            continue;
+          }
+        }
+        break;
+      }
+      msg((char_u *)str + start);
+    }
+    if (str[len - 1] == NUL) {  // Last was newline
+      msg((char_u *)"");
+    }
+  }
+  ga_clear(&msg_ga);
+  return 0;
+nlua_print_error:
+  emsgf(_("E5114: Error while converting print argument #%i: %.*s"),
+        curargidx, errmsg_len, errmsg);
+  ga_clear(&msg_ga);
+  lua_pop(lstate, lua_gettop(lstate));
+  return 0;
+}
+
+/// debug.debug implementation: interaction with user while debugging
+///
+/// @param  lstate  Lua interpreter state.
+int nlua_debug(lua_State *lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  const typval_T input_args[] = {
+    {
+      .v_lock = VAR_FIXED,
+      .v_type = VAR_STRING,
+      .vval.v_string = (char_u *)"lua_debug> ",
+    },
+    {
+      .v_type = VAR_UNKNOWN,
+    },
+  };
+  for (;;) {
+    lua_settop(lstate, 0);
+    typval_T input;
+    get_user_input(input_args, &input, false);
+    msg_putchar('\n');  // Avoid outputting on input line.
+    if (input.v_type != VAR_STRING
+        || input.vval.v_string == NULL
+        || *input.vval.v_string == NUL
+        || STRCMP(input.vval.v_string, "cont") == 0) {
+      tv_clear(&input);
+      return 0;
+    }
+    if (luaL_loadbuffer(lstate, (const char *)input.vval.v_string,
+                        STRLEN(input.vval.v_string), "=(debug command)")) {
+      nlua_error(lstate, _("E5115: Error while loading debug string: %.*s"));
+    }
+    tv_clear(&input);
+    if (lua_pcall(lstate, 0, 0, 0)) {
+      nlua_error(lstate, _("E5116: Error while calling debug string: %.*s"));
+    }
+  }
+  return 0;
+}
+
+/// Evaluate lua string
+///
+/// Used for luaeval().
+///
+/// @param[in]  str  String to execute.
+/// @param[in]  arg  Second argument to `luaeval()`.
+/// @param[out]  ret_tv  Location where result will be saved.
+///
+/// @return Result of the execution.
+void executor_eval_lua(const String str, typval_T *const arg,
+                       typval_T *const ret_tv)
+  FUNC_ATTR_NONNULL_ALL
+{
+  if (global_lstate == NULL) {
+    global_lstate = init_lua();
+  }
+
+  NLUA_CALL_C_FUNCTION_3(global_lstate, nlua_eval_lua_string, 0,
```

Comment:
I am not sure. Vim does nearly the same thing, except that it for some reasons first saves functions into a registry and I do not do that (have no idea why this is needed, as well as the registry itself, unlike some ideas for why such indirection is needed). Possible reasonings are

1. Error reporting: additional call stack level to indicate that Vim is using a C function (appears in `debug.traceback()` as `[C]: ?` at the bottom, both in Vim and Neovim).
2. It is for some reason needed to separate a stack.
3. It is needed for error handling: you will not get a jump point without lua_call. Rather unsure that this is the case: jump points should be created by lua_pcall and not by lua_call.

I would rather not change that unless the author of if_lua explains why he created bindings like this.